ddtrace/tracer/tracer.go

```diff
@@ -192,27 +197,27 @@ func (t *tracer) worker(tick <-chan time.Time) {
 
 	for {
 		select {
-		case trace := <-t.payloadChan:
-			t.pushPayload(trace)
+		case trace := <-t.out:
+			t.traceWriter.add(trace)
 
 		case <-tick:
 			t.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:scheduled"}, 1)
-			t.flush()
+			t.traceWriter.flush()
 
 		case <-t.stop:
 		loop:
 			// the loop ensures that the payload channel is fully drained
 			// before the final flush to ensure no traces are lost (see #526)
 			for {
 				select {
-				case trace := <-t.payloadChan:
-					t.pushPayload(trace)
+				case trace := <-t.out:
+					t.traceWriter.add(trace)
 				default:
 					break loop
 				}
 			}
 			t.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:shutdown"}, 1)
-			t.flush()
+			t.traceWriter.flush()
```

Comment:
Do we need this given that `Stop` calls `traceWriter.stop()` which could also flush?

Reply:
Why can't we make `stop()` flush before waiting?

Reply:
We can.
It would remove this one `flush()` call and add one in each `stop()` method. Since this is behavior that every writer should have, I chose to do it here rather than in each writer, but I think having `stop()` perform a `flush()` makes the writer interface more intuitive.

Reply:
Thanks!

> [...] but I think having stop() perform a flush() makes the writer interface more intuitive.

Not only that, but I think you'd expect that if something's shutting down gracefully, it would also clean up and flush whatever's left. That's how our tracer's behaving too.